internal/api/domain_remap.go

```diff
@@ -0,0 +1,106 @@
+/*******************************************************************************
+*
+* Copyright 2021 SAP SE
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You should have received a copy of the License along with this
+* program. If not, you may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*
+*******************************************************************************/
+
+package api
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/sapcc/keppel/internal/keppel"
+)
+
+//AddDomainRemapMiddleware adds to the handler chain a middleware that implements domain remapping, i.e. rewriting of URLs like
+//
+//    https://<account>.<public-hostname>/v2/<path>
+//
+//into
+//
+//    https://<public-hostname>/v2/<account>/<path>
+//
+func AddDomainRemapMiddleware(cfg keppel.Configuration, h http.Handler) http.Handler {
+	return domainRemap{cfg, h}
+}
+
+type domainRemap struct {
+	cfg  keppel.Configuration
+	next http.Handler
+}
+
+//ServeHTTP implements the http.Handler interface.
+func (h domainRemap) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	//if request does not use domain remapping, forward to the next handler unchanged
+	rewrittenURL, matches := h.tryRewriteURL(keppel.OriginalRequestURL(r))
+	if !matches {
+		h.next.ServeHTTP(w, r)
+		return
+	}
```

Comment:
I think we should differentiate between no domain remapping and wrong domain remapping before people start to use their subdomains and we can't easily remove it anymore.  eg ``https://myaccount.registry.example.com/v2/myaccount/myrepo/manifests/latest`` should error differently and shouldn't be passed through.

Reply:
Your example URL is legal though. Repos can have a path, so in your case the account would be `myaccount` and the repo would be `myaccount/myrepo`, corresponding to `https://registry.example.com/v2/myaccount/myaccount/myrepo/manifests/latest`. It's not unreasonable to assume that such scenarios would arise in real-life usage, e.g. if the account is named after a team's main product.

In fact, I did a small search for `SELECT * FROM repos WHERE name LIKE CONCAT(account_name, '/%')` and found several such repos in prod.

Reply:
and what would happen if ``myaccount.registry.example.com/myaccount/myrepo/manifests/latest`` was used?

Reply:
Good point, I've added some extra logic to explicitly forbid anything that's not `/v2/...`.